### DISCUSSION TITLE
"Source file already exists" with ECJ (and Maven)

### DISCUSSION THREAD

**Discussion by famod** [mapstruct/mapstruct#2920] [Q&A]
I'm getting:
```
Internal error in the mapping processor: java.lang.RuntimeException: javax.annotation.processing.FilerException: Source file already exists : foo.MyMapperImpl
...
Caused by: javax.annotation.processing.FilerException: Source file already exists : foo.MyMapperImpl
    at org.eclipse.jdt.internal.compiler.apt.dispatch.BatchFilerImpl.createSourceFile(BatchFilerImpl.java:149)
    at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.createSourceFile(MapperRenderingProcessor.java:56)
    ... 34 more
```

when using `maven-compiler-plugin` with `plexus-compiler-eclipse`, **but only when building _without_ `clean` _and_ having changed at least one file (doesn't need to be a mapper though!)**.

Any hints? Thanks!

PS: This doesn't happen if using javac.
PPS: There is a [bug](https://github.com/codehaus-plexus/plexus-compiler/issues/229) in `plexus-compiler-eclipse` that passes duplicate mapper sources to ECJ, but the above is with that bug being fixed locally. (otherwise it would shadow the above issue)

**Comment by Zegveld** [comment]
This error message also happens when you have something else that adds the mapstruct generated sources to the source list in maven. Do you have any other maven plugins that modify the compilation path?

For example having the maven-helper-plugin point to that same folder, or another annotation processor that uses the root folder of where mapstruct creates its files.

**Reply by famod** [reply]
Well, there is also lombok in the mix and openapi-generator.

After reading https://bugs.eclipse.org/bugs/show_bug.cgi?id=547010 and especially https://bugs.eclipse.org/bugs/show_bug.cgi?id=542090 I'm under the impression that it's probably a flaw in ECJ (also given that the very same setup works with javac).
